src/sleep/sleep.c

```diff
@@ -264,77 +264,110 @@ static int execute(
 
 static int execute_s2h(const SleepConfig *sleep_config) {
         _cleanup_hashmap_free_ Hashmap *last_capacity = NULL, *current_capacity = NULL;
-        int r;
+        int wakeup_type_bit, r;
 
         assert(sleep_config);
 
-        while (battery_is_low() == 0) {
-                _cleanup_close_ int tfd = -1;
-                struct itimerspec ts = {};
-                usec_t suspend_interval = sleep_config->hibernate_delay_sec, before_timestamp = 0, after_timestamp = 0, total_suspend_interval;
-                bool woken_by_timer;
-
-                tfd = timerfd_create(CLOCK_BOOTTIME_ALARM, TFD_NONBLOCK|TFD_CLOEXEC);
-                if (tfd < 0)
-                        return log_error_errno(errno, "Error creating timerfd: %m");
-
-                /* Store current battery capacity and current time before suspension */
-                r = fetch_batteries_capacity_by_name(&last_capacity);
-                if (r >= 0)
-                        before_timestamp = now(CLOCK_BOOTTIME);
-                else if (r == -ENOENT)
-                        /* In case of no battery, system suspend interval will be set to HibernateDelaySec=. */
-                        log_debug_errno(r, "Suspend Interval value set to %s: %m", FORMAT_TIMESPAN(suspend_interval, USEC_PER_SEC));
-                else
-                        return log_error_errno(r, "Error fetching battery capacity percentage: %m");
-
-                r = get_total_suspend_interval(last_capacity, &total_suspend_interval);
-                if (r < 0)
-                        log_debug_errno(r, "Failed to estimate suspend interval using previous discharge rate, ignoring: %m");
-                else
-                        suspend_interval = total_suspend_interval;
-
-                log_debug("Set timerfd wake alarm for %s", FORMAT_TIMESPAN(suspend_interval, USEC_PER_SEC));
-                /* Wake alarm for system with or without battery to hibernate or estimate discharge rate whichever is applicable */
-                timespec_store(&ts.it_value, suspend_interval);
+        r = check_wakeup_type(&wakeup_type_bit);
+        if (r < 0)
+                log_debug_errno(r, "Failed to fetch wakeup-type bit: %m");
 
-                if (timerfd_settime(tfd, 0, &ts, NULL) < 0)
-                        return log_error_errno(errno, "Error setting battery estimate timer: %m");
+        r = battery_trip_point_alarm_exists();
+        if (r < 0)
+                log_debug_errno(r, "Failed to check acpi_btp support enabled or not: %m");
 
+        if (r > 0 && wakeup_type_bit > 0) {
+                log_debug("Attempting to suspend...");
                 r = execute(sleep_config, SLEEP_SUSPEND, NULL);
                 if (r < 0)
                         return r;
 
-                r = fd_wait_for_event(tfd, POLLIN, 0);
+                r = check_wakeup_type(&wakeup_type_bit);
```

Comment:
Hm? Why this is called again?

Reply:
After the indefinite suspend, if the system wakes up, there could 2 possibilities
- manual wakeup
- APM Timer

If it is APM timer wakeup for which this wakeup bit will be set to 3 as per Table 12 [Line 973]
https://www.dmtf.org/sites/default/files/standards/documents/DSP0134_3.4.0.pdf , it means the battery is Low. In that case the system should hibernate.